libs/platform/src/lib/components/form/combobox/commons/base-combobox.ts

```diff
@@ -0,0 +1,749 @@
+        protected _menuKeyboardService: MenuKeyboardService,
+        protected _comboboxConfig: ComboboxConfig
+    ) {
+        super(_cd, ngControl, ngForm);
+    }
+
+    /** @hidden */
+    ngAfterViewInit(): void {
+        this._setupKeyboardService();
+        this._initWindowResize();
+        this._assignCustomTemplates();
+        super.ngAfterViewInit();
+
+        if (this.mobile) {
+            this._setUpMobileMode();
+        }
+    }
+
+    /** @hidden */
+    ngOnDestroy(): void {
+        super.ngOnDestroy();
+
+        if (isDataSource(this.dataSource)) {
+            (this.dataSource as ComboBoxDataSource<any>).close();
+        }
+
+        if (this._dsSubscription) {
+            this._dsSubscription.unsubscribe();
+        }
+    }
+
+    /** @hidden
+     * Method to emit change event
+     */
+    abstract emitChangeEvent<K>(value: K): void;
+
+    /** @hidden
+     * Define is this item selected
+     */
+    abstract isSelectedOptionItem(selectedItem: OptionItem): boolean;
+
+    /** @hidden
+     * Emit select OptionItem
+     * */
+    abstract selectOptionItem(item: OptionItem): void;
+
+    /** @hidden
+     * Define value as selected
+     * */
+    abstract setAsSelected(item: any[]): void;
+
+    /** write value for ControlValueAccessor */
+    writeValue(value: any): void {
+        const selectedItems = Array.isArray(value) ? value : [value];
+        this.setAsSelected(selectedItems);
+        super.writeValue(value);
+    }
+
+    /** @hidden
+     * Close list
+     * */
+    close(event: MouseEvent = null, forceClose: boolean = false): void {
+        if (event) {
+            const target = event.target as HTMLInputElement;
+            if (target && target.id === this.id) {
+                return;
+            }
+        }
+
+        if (this.isOpen && (forceClose || this.canClose)) {
+            this.isOpen = false;
+            this._cd.markForCheck();
+            this.onTouched();
+        }
+    }
+
+    /** @hidden */
+    searchTermChanged(text: string = this.inputText): void {
+        const map = new Map();
+        map.set('query', text);
+        map.set('limit', 12);
+
+        this.ds.match(map);
+    }
+
+    /** @hidden */
+    showList(isOpen: boolean): void {
+        /** Reset displayed values on every mobile open */
+        if (this.mobile && !this.isOpen) {
+            this.searchTermChanged('');
+        }
+
+        if (this.isOpen !== isOpen) {
+            this.isOpen = isOpen;
+            this.onTouched();
+            this.openChange.next(isOpen);
+        }
+
+        this._cd.detectChanges();
+    }
+
+    /** @hidden */
+    handleOptionItem(value: any): void {
+        if (value) {
+            this.selectOptionItem(value);
+        }
+    }
+
+    /**
+     * Handle Keydown Actions
+     * @hidden
+     */
+    onListKeydownHandler(event: KeyboardEvent): void {
+        const index: number = this.listItems.toArray().findIndex(
```

Comment:
We can optimize a little bit here and run `.toArray()` once 😛

Reply:
fixed